File: server2/src/domain/games/game.py
from typing import List
from src.domain.id import Id
from src.domain.players.player import Player


class GameId(Id):
    def __init__(self, value: str):
        self._value = value

    @staticmethod
    def from_host_username(host_username: str):
        return GameId(host_username)


class Game:
    def __init__(self, host_username: str, nb_players: int, world: str):
        self.__id = GameId.from_host_username(host_username)
        self.__nb_players = nb_players
        self.__world = world

        self.__players: List[Player] = []
        self.__is_available = True

    @property
    def id(self) -> GameId:
        return self.__id

    @property
    def is_available(self) -> bool:
        return self.__is_available

    @property
    def players_remaining(self):
        return self.__nb_players - len(self.__players)

    def __is_full(self) -> bool:
        return len(self.__players) == self.__nb_players

    def add_player(self, player):
        self.__players.append(player)

        if self.__is_full():
            self.__is_available = False
            for player in self.__players:
                # event_emitter.send(player.sid, )
                pass

    def remove_player(self, player):
        pass

File: new-server/src/game/game.ts
import { ArgumentError } from '../shared/exceptions'
import { GameId } from './gameId'
import { Player } from '../player/player'
import { findIndexByEquality } from '../shared/arrayutils'
import { Equalable } from '../shared/equalable'

export class Game implements Equalable<Game> {
  readonly id: GameId
  private _isAvailable = true
  private nbPlayers: number
  private players: Array<Player> = []

  constructor(hostUsername: string, nbPlayers: number, private world: string) {
    if (nbPlayers !== Math.round(nbPlayers)) {
      throw new ArgumentError("'nb_players' should be an integer")
    }
    if (nbPlayers < 1 || nbPlayers > 4) {
      throw new ArgumentError("'nb_players' should be between 1 and 4")
    }
    this.id = GameId.fromHostUsername(hostUsername)
    this.nbPlayers = nbPlayers
  }

  playersRemaining(): number {
    return this.nbPlayers - this.players.length
  }

  addPlayer(player: Player): void {
    if (!this._isAvailable) {
      throw new ArgumentError('the game has no room for new players')
    }

    player.setGameId(this.id)
    this.players.push(player)

    if (this.playersRemaining() === 0) {
      this._isAvailable = false
      // TODO send start event with this.world
    }
  }

  removePlayer(player: Player): void {
    const index = findIndexByEquality(this.players, player)
    if (index === -1) {
      throw new ArgumentError("can't remove player: not present")
    }

    this.players.splice(index, 1)
  }

  isAvailable(): boolean {
    return this._isAvailable
  }

  equals(other: Game): boolean {
    return this.id.equals(other.id)
  }
}
